File: sdks/typescript/src/types/graph.ts
/**
 * Graph, Entity, and Relationship types.
 *
 * @module types/graph
 * @see edgequake/crates/edgequake-api/src/handlers/graph_types.rs
 * @see edgequake/crates/edgequake-api/src/handlers/entities_types.rs
 * @see edgequake/crates/edgequake-api/src/handlers/relationships_types.rs
 */

import type { Timestamp } from "./common.js";

// ── Graph ─────────────────────────────────────────────────────

export interface GraphQuery {
  limit?: number;
  labels?: string[];
  search?: string;
}

export interface GraphNode {
  id: string;
  label: string;
  properties?: Record<string, unknown>;
  degree?: number;
}

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
  label: string;
  weight?: number;
  properties?: Record<string, unknown>;
}

export interface GraphResponse {
  nodes: GraphNode[];
  edges: GraphEdge[];
  total_nodes: number;
  total_edges: number;
}

export type GraphStreamEvent =
  | { type: "node"; node: GraphNode }
  | { type: "edge"; edge: GraphEdge }
  | { type: "done"; total_nodes: number; total_edges: number };

export interface SearchNodesResponse {
  nodes: GraphNode[];
  total: number;
}

export interface SearchLabelsResponse {
  labels: Array<{ label: string; count: number }>;
}

export interface PopularLabelsResponse {
  labels: Array<{ label: string; count: number }>;
}

export interface DegreesBatchResponse {
  degrees: Record<string, number>;
}

export interface SearchParams {
  limit?: number;
  offset?: number;
}

// ── Entities ──────────────────────────────────────────────────

export interface ListEntitiesQuery {
  page?: number;
  per_page?: number;
  search?: string;
  label?: string;
}

export interface EntitiesListResponse {
  entities: EntityInfo[];
  total: number;
  page: number;
  per_page: number;
}

export interface EntityInfo {
  name: string;
  label: string;
  description?: string;
  source_count?: number;
  created_at?: Timestamp;
}

export interface EntityResponse {
  name: string;
  label: string;
  description?: string;
  properties?: Record<string, unknown>;
  source_documents?: string[];
}

export interface CreateEntityRequest {
  name: string;
  label: string;
  description?: string;
  properties?: Record<string, unknown>;
}

export interface UpdateEntityRequest {
  description?: string;
  properties?: Record<string, unknown>;
}

export interface MergeEntitiesRequest {
  source_entity: string;
  target_entity: string;
  strategy?: "keep_target" | "keep_source" | "merge";
}

export interface MergeEntitiesResponse {
  merged_entity: string;
  removed_entity: string;
  relationships_updated: number;
}

export interface NeighborhoodResponse {
  center: EntityResponse;
  neighbors: Array<{
    entity: EntityResponse;
    relationship: string;
    direction: "incoming" | "outgoing";
  }>;
  depth: number;
}

// ── Relationships ─────────────────────────────────────────────

export interface ListRelationshipsQuery {
  page?: number;
  per_page?: number;
  source?: string;
  target?: string;
  label?: string;
}

export interface RelationshipsListResponse {
  relationships: RelationshipInfo[];
  total: number;
  page: number;
  per_page: number;
}

export interface RelationshipInfo {
  id: string;
  source: string;
  target: string;
  label: string;
  weight?: number;
  description?: string;
  created_at?: Timestamp;
}

export interface RelationshipResponse extends RelationshipInfo {
  properties?: Record<string, unknown>;
  source_documents?: string[];
}

export interface CreateRelationshipRequest {
  source: string;
  target: string;
  label: string;
  weight?: number;
  description?: string;
  properties?: Record<string, unknown>;
}

export interface UpdateRelationshipRequest {
  weight?: number;
  description?: string;
  properties?: Record<string, unknown>;
}

// ── Query Helpers ─────────────────────────────────────────────

export interface SearchNodesQuery {
  q?: string;
  limit?: number;
}

export interface SearchLabelsQuery {
  q?: string;
  limit?: number;
}

export interface DegreeBatchRequest {
  node_ids: string[];
}

export interface DegreeBatchResponse {
  degrees: Record<string, number>;
}

// ── Type Aliases for resource usage ───────────────────────────

/** Entity detail (alias for EntityResponse). */
export type EntityDetail = EntityResponse;

/** Entity neighborhood (alias for NeighborhoodResponse). */
export type EntityNeighborhood = NeighborhoodResponse;

/** Relationship detail (alias for RelationshipResponse). */
export type RelationshipDetail = RelationshipResponse;

File: sdks/rust/src/types/graph.rs
//! Graph-related types.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// A node in the knowledge graph.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GraphNode {
    pub id: String,
    pub label: String,
    #[serde(default)]
    pub node_type: Option<String>,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub properties: Option<HashMap<String, serde_json::Value>>,
    #[serde(default)]
    pub degree: Option<u32>,
}

/// An edge in the knowledge graph.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GraphEdge {
    pub source: String,
    pub target: String,
    #[serde(default)]
    pub edge_type: Option<String>,
    #[serde(default)]
    pub weight: Option<f64>,
    #[serde(default)]
    pub properties: Option<HashMap<String, serde_json::Value>>,
}

/// Full graph response.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GraphResponse {
    #[serde(default)]
    pub nodes: Vec<GraphNode>,
    #[serde(default)]
    pub edges: Vec<GraphEdge>,
    #[serde(default)]
    pub total_nodes: Option<u32>,
    #[serde(default)]
    pub total_edges: Option<u32>,
}

/// Search nodes response.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct SearchNodesResponse {
    #[serde(default)]
    pub nodes: Vec<GraphNode>,
    #[serde(default)]
    pub edges: Vec<GraphEdge>,
    #[serde(default)]
    pub total_matches: Option<u32>,
}

/// Entity.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Entity {
    pub name: String,
    #[serde(default)]
    pub entity_type: Option<String>,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub properties: Option<HashMap<String, serde_json::Value>>,
    #[serde(default)]
    pub degree: Option<u32>,
    #[serde(default)]
    pub created_at: Option<String>,
}

/// Request to create an entity.
#[derive(Debug, Clone, Serialize)]
pub struct CreateEntityRequest {
    pub name: String,
    pub entity_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub properties: Option<HashMap<String, serde_json::Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_id: Option<String>,
}

/// Entity exists check.
#[derive(Debug, Clone, Deserialize)]
pub struct EntityExistsResponse {
    pub exists: bool,
    #[serde(default)]
    pub entity_name: Option<String>,
}

/// Merge entities response.
#[derive(Debug, Clone, Deserialize)]
pub struct MergeEntitiesResponse {
    #[serde(default)]
    pub merged_entity: Option<Entity>,
    #[serde(default)]
    pub merged_count: u32,
    #[serde(default)]
    pub message: Option<String>,
}

/// Merge entities request.
#[derive(Debug, Clone, Serialize)]
pub struct MergeEntitiesRequest {
    pub source: String,
    pub target: String,
}

/// Neighborhood response.
#[derive(Debug, Clone, Deserialize)]
pub struct NeighborhoodResponse {
    #[serde(default)]
    pub center: Option<Entity>,
    #[serde(default)]
    pub nodes: Vec<GraphNode>,
    #[serde(default)]
    pub edges: Vec<GraphEdge>,
    #[serde(default = "default_depth")]
    pub depth: u32,
}

fn default_depth() -> u32 {
    1
}

/// Relationship.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Relationship {
    #[serde(default)]
    pub id: Option<String>,
    pub source: String,
    pub target: String,
    #[serde(default)]
    pub relationship_type: Option<String>,
    #[serde(default)]
    pub weight: Option<f64>,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub properties: Option<HashMap<String, serde_json::Value>>,
}

/// Request to create a relationship.
#[derive(Debug, Clone, Serialize)]
pub struct CreateRelationshipRequest {
    pub source: String,
    pub target: String,
    pub relationship_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub weight: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

/// Degrees batch response.
#[derive(Debug, Clone, Deserialize)]
pub struct DegreesBatchResponse {
    #[serde(default)]
    pub degrees: HashMap<String, u32>,
}
